machine-learning/sagemaker-edge-manager/artifacts/com.greengrass.SageMakerEdgeManager.ObjectDetection/1.0.0/object_detection/prediction_utils.py

```diff
@@ -0,0 +1,169 @@
+    img_data = transform_image(image)
+    img_data = cv2.resize(img_data, config_utils.SHAPE)
+    img_data = img_data[:, :, (2, 1, 0)].astype(np.float32)
+    img_data -= np.array([123, 117, 104])
+    img_data = np.transpose(np.array(img_data), (2, 0, 1))
+    img_data = np.expand_dims(img_data, axis=0)
+    predict(img_data)
+
+
+def load_image(image_path):
+    r"""
+    Validates the image type irrespective of its case. For eg. both .PNG and .png are valid image types.
+    Also, accepts numpy array images.
+
+    :param image_path: path of the image on the device.
+    :return: a numpy array of shape (1, input_shape_x, input_shape_y, no_of_channels)
+    """
+    # Case insenstive check of the image type.
+    img_lower = image_path.lower()
+    if (
+        img_lower.endswith(
+            ".jpg",
+            -4,
+        )
+        or img_lower.endswith(
+            ".png",
+            -4,
+        )
+        or img_lower.endswith(
+            ".jpeg",
+            -5,
+        )
+    ):
+        try:
+            image_data = cv2.imread(image_path)
+        except Exception as e:
+            config_utils.logger.error(
+                "Unable to read the image at: {}. Error: {}".format(image_path, e)
+            )
+            exit(1)
+    elif img_lower.endswith(
+        ".npy",
+        -4,
+    ):
+        image_data = np.load(image_path)
+    else:
+        config_utils.logger.error("Images of format jpg, jpeg, png and npy are only supported.")
+        exit(1)
+    return image_data
+
+
+def load_model(name, url):
+    load_request = LoadModelRequest()
+    load_request.url = url
+    load_request.name = name
+    config_utils.agent_client.LoadModel(load_request)
+
+
+def unload_model(name):
+    unload_request = UnLoadModelRequest()
+    unload_request.name = name
+    config_utils.agent_client.UnLoadModel(unload_request)
+
+
+def predict(image_data):
+    PAYLOAD = {}
+    PAYLOAD["timestamp"] = str(datetime.now())
+    PAYLOAD["inference-type"] = "object-detection"
+    PAYLOAD["inference-description"] = "Top {} predictions with score {} or above ".format(
+        config_utils.MAX_NO_OF_RESULTS, config_utils.SCORE_THRESHOLD
+    )
+    PAYLOAD["inference-results"] = []
+
+    request = PredictRequest(
+        name=config_utils.MODEL_NAME,
+        tensors=[
+            Tensor(
+                tensor_metadata=TensorMetadata(
+                    name=config_utils.tensor_name,
+                    data_type=5,
+                    shape=config_utils.tensor_shape,
+                ),
+                byte_data=image_data.tobytes(),
+            )
+        ],
+    )
+    response = config_utils.agent_client.Predict(request)
+
+    detections = []
+
+    for t in response.tensors:
+        deserialized_bytes = np.frombuffer(t.byte_data, dtype=np.float32)
+        detections.append(np.asarray(deserialized_bytes).tolist())
+
+    # Get the bounding boxes
+    new_list = []
+    for index, item in enumerate(detections[2]):
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

To create a `list`, try to use `list` comprehension instead of a loop. List comprehension is the preferred way to make a list using Python, and it&#39;s simpler and easier to understand than using a loop.

[Learn more](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions)